content/struct-classes.md

```diff
@@ -0,0 +1,502 @@
+  def this(name: String, age: Int, email: Option[String]): User = this(name, age)
+  def withName(newName: String): User = copy(name = newName)
+  def withAge(newAge: Int): User = copy(age = newAge)
+  // "email" methods, for compatibility
+  def email: Option[String] = None
+  def withEmail(newEmail: Option[String]): User = this
+~~~
+
+Note that the fact that we explicitly define the transformation methods (here,
+`withName`, `withAge`, etc.) may be seen as a lack of "language support". On the
+contrary, this gives more flexibility to the developers to define custom transformation
+methods. For instance, the following snippet shows two possibilities of transformation
+methods to set the value of the `email` field:
+
+~~~ scala
+// 1st solution, as above
+def withEmail(newEmail: Option[String]): User
+// 2nd solution
+def withEmail(newEmail: String): User
+def withoutEmail: User
+~~~
+
+Here, in the 2nd solution, calling `withEmail` would always set the email to
+`Some(newEmail)`, and calling `withoutEmail` would set it to `None`. We can imagine
+similar convenience methods for fields containing collections of values.
+
+Internally, a `struct class` is implemented in a similar way as a `case class`: the
+compiler synthesizes the `private` method `copy`, and overrides the default implementations
+of `toString`, `hashCode`, and `equals`. For reference, here is what the definition
+of the first version of `User` "desugars" to:
+
+~~~ scala
+import scala.util.hashing.MurmurHash3
+
+class User(val name: String, val age: Int) extends Serializable:
+  private def copy(name: String = name, age: Int = age): User = User(name, age)
+  override def toString() = s"User($name, $age)"
+  override def hashCode() =
+    MurmurHash3.finalizeHash(
+      MurmurHash3.mix(MurmurHash3.mix(MurmurHash3.productSeed, name), age),
+      2
+    )
+  override def equals(that: Any): Boolean =
+    that match
+      case user: User => user.name == name && user.age == age
+      case _ => false
+
+  def withName(newName: String): User = copy(name = newName)
+  def withAge(newAge: Int): User = copy(age = newAge)
+end User
+~~~
+
+### Specification
+
+A struct class is a `class` definition prefixed with `struct`:
+
+~~~ text
+TmplDef ::= 'struct' 'class' ClassDef
+~~~
+
+A struct class is required to have a parameter list that is not contextual. The
+formal parameters in the first parameter list are called _elements_ and are treated
+specially. A `val` prefix is implicitly added to such a parameter, unless the parameter
+already carries a `val` or `var` modifier. Hence, an accessor definition for the
+parameter is generated.
+
+A struct class definition `c[tps](ps_1)...(ps_n)` with type parameters `tps` and
+value parameters `ps` is handled as follows. A `private` method named `copy` is
+implicitly added to the class definition unless the class already has a member
+(directly defined or inherited) with that name, or the class has a repeated parameter.
```

Comment:
Is it just the name `copy` used to decide generation? Or would other parts of an existing function signature, such as the return type used to determine if it needs to generate a copy method?

Reply:
I copied and adapted the specification of case classes from [here](https://scala-lang.org/files/archive/spec/2.13/05-classes-and-objects.html#case-classes). The reason is that I want the same schema to be applied both to struct classes and case classes.

In that specification, only the name is used, but not the signature. I’ve noticed that the Scala 2 compiler does not synthesize a `copy` method if I already define one (even if it’s signature does not match the case class fields). However, I’ve just noticed that the Scala 3 compiler does not behave the same here: if you add a method `copy`, the compiler adds another one as an overload. I’ve created https://github.com/lampepfl/dotty/discussions/16309 to clarify whether this is expected or not.